README.md
# ModuleCompiler

[![Build Status](https://travis-ci.org/daxingplay/kmc.svg?branch=master)](https://travis-ci.org/daxingplay/kmc)

[![NPM version](https://badge.fury.io/js/kmc.png)](http://badge.fury.io/js/kmc)

## 简介

KISSY Module Compiler（kmc）是一个基于NodeJS的KISSY模块打包工具，目前适用于KISSY 1.2+的代码打包

如果您有任何问题，请邮件或者来往（daxingplay）上联系我，阿里巴巴厂内用户还可以联系紫英。

## 特点

- 支持GruntJS，参见[grunt-kmc](https://github.com/daxingplay/grunt-kmc).
- 基于NodeJS，相比于KISSY自带的Java工具，打包快速
- 参照浏览器端的KISSY的config进行配置，无需额外知识，只需要改一下包路径即能快速打包
- 支持混合编码打包，不同的包可以使用不同的编码
- 支持GBK输出
- 支持KISSY 1.3的自动combo功能，可以生成依赖关系文件
- 提供底层依赖分析接口，方便集成到其他工具当中
- 支持map功能，可以使用正则自由替换输出的模块名

## 版本说明

- 0.0.7版本适用于KISSY 1.2、1.3的打包，目前已经在淘宝多个业务广泛使用，单纯打包没有任何问题，但是不具备依赖分析生成功能，此版本已经不再维护，推荐使用新版本。
- 1.0.0版本开始支持KISSY 1.3的自动combo功能，推荐使用

## 使用

### 安装
    npm install kmc

or

    git clone git://github.com/daxingplay/kmc.git

**注意**: 新版本的KISSY Module Compiler的npm包已经更名为kmc，如果使用老版本（此版已经不再维护），请`npm install module-compiler`

### 编写你的打包脚本

```js
var kmc = require('kmc');

// 这里和KISSY.config一样，先配置包
kmc.config({
    packages: [{
        'name': 'sh',
        'path': '这里建议写绝对路径，即sh这个包所在的目录',
        'charset': 'gbk'
    }]
});

// 将xxx.js打包为xxx.combine.js，输出编码为GBK
kmc.build('xxx.js', 'xxx.combine.js', 'gbk');

// 用node执行你这个打包脚本就ok啦～
```

### 高级使用指南

```js
var kmc = require('kmc');

kmc.config({
    // 和KISSY一样，可以配置多个包
    packages: [{
        'name': 'app1',
        'path': 'app1这个包所在目录的绝对路径',
        // 这里是指app1这个包中的文件的编码，同一个包内的编码请保持一致
        'charset': 'gbk'
    }, {
        'name': 'app2',
        'path': 'app2这个包所在目录的绝对路径',
        // 这里是指app2这个包源码的编码
        'charset': 'utf-8'
    }],
    // 可以设置哪些模块不打包进来。注意，这里exclude的是具体的模块名，支持正则
    exclude: ['base', 'event'],
    // 如果是对一个目录下的所有文件进行打包，可以设置哪些文件不打包进来，支持正则。注意和上面的exclude的配置的区别。
    ignoreFiles: ['.combo.js', '-min.js'],
    // 输出的文件名后缀，不带.js，比如打包后你想输出为xxx.combine.js，那么这里就配置为：.combine
    suffix: '',
    // 类似于KISSY的map方法，可以自己定义把模块名中的路径进行替换
    map: [
        // 这样配置的话，那么，如果原先输出的app1的模块名中含有app1/2.0/字样的话，就会被替换成app1/19891014/
        ['app1/2.0/', 'app1/19891014/']
    ],
    // 这里设置的是最后打包出来的文件的编码，默认UTF-8，这里的设置相当于是全局设置，下面build中的设置是针对单一打包实例的
    charset: 'gbk'
});

/**
 * 打包一个文件/目录
 * @param src {String} 源文件/目录的绝对路径.
 * @param dest {String} 打包出来的文件/目录的路径.
 * @param outputCharset {String} 输出编码，这里的设置会覆盖config.charset中的设置，默认UTF-8
 * @return {Object} 打包出来的文件信息
 */
kmc.build({
    src: 'xxx.js',
    dest: 'xxx.combine.js',
    outputCharset: 'gbk'
});
```

更详细的文档，请参见[wiki](https://github.com/daxingplay/kmc/wiki)。

### API汇总

#### 配置

配置包，返回当前所有配置信息。如果不带参数，直接返回当前所有配置信息。

```js
kmc.config(cfg);
```

#### 分析依赖

只分析该文件依赖，不打包。

```js
kmc.analyze(src);
```

#### 打包

打包函数，具体见wiki

```js
kmc.build({
	src: '\xxx\xxx\src\a.js', // 需要打包的文件
	dest: '\xxx\xxx\build\a.js', // 打包后文件输出的路径
	outputCharset: 'gbk', // 所有的输出文件编码，包括依赖文件
	depPath: '\xxx\xxx\dep.js' // 依赖文件的路径
});
```


#### 生成依赖

不打包，只生成KISSY 1.3的自动combo依赖文件

```js
kmc.combo({
	src: '', // 需要打包or生成依赖关系的文件
	dest: '', // 模块最后输出的路径（fixModuleName === true的时候必须配置这个选项，否则源文件的内容会被修改）
	depPath: '', // 依赖文件的输出路径
	depCharset: '', // 依赖文件的编码
	fixModuleName: '', // 是否给文件添加模块名，默认为true
	showFullResult: '' // 函数的return结果会以对象的形式输出丰富的信息
});
```

#### 获取依赖

```js
kmc.info(inputFile) // => {modules: object, dependencies: object}
```

#### 清空配置

可以清空config中的设置。因为ModuleCompiler是单例运行，所以如果出现一些特别情况，可以在config前执行clean方法清空之前的配置。

```js
kmc.clean();
```


## CHANGELOG

[版本更新记录](https://github.com/daxingplay/kmc/blob/master/HISTORY.md)

## License
遵守 "MIT"：https://github.com/daxingplay/kmc/blob/master/LICENSE.md 协议

HISTORY.md
## 2015.03.19, Version 1.2.0(Stable)

- 修复主文件被package干扰导致打包为空的bug

## 2014.11.19, Version 1.0.42(Stable)

- merge pull request[#49](https://github.com/daxingplay/kmc/pull/49)

## 2014.07.08, Version 1.0.41(Stable)

### bugfix

- 修复[issue #13](https://github.com/daxingplay/grunt-kmc/issues/13)

## 2014.06.20, Version 1.0.40(Stable)

### bugfix

- 修复module的默认编码问题

## 2014.06.20, Version 1.0.39(Stable)

### bugfix

- 修复combo后的dep文件的模块名多一个引号的bug

## 2014.04.21, Version 1.0.37(Stable)

### bugfix

- 修复windows下模块名斜线错误

## 2014.04.14, Version 1.0.36(Stable)

### bugfix

- issue #44: 修复包名.js文件模块名错误

## 2014.02.20, Version 1.0.34(Stable)

### bugfix

- 如果模块没有找到，仍然加到依赖关系中，模块的包默认为unknown

## 2014.01.18, Version 1.0.33(Stable)

### bugfix

- 如果require用错了，暂时不抛出异常，配置了silent: false后可以看到console的输出

## 2013.12.23, Version 1.0.31(Stable)

### bugfix

- 修复一种代码书写的方式造成的依赖分析问题，参考test/src/kissy/kissy-extend.js

## 2013.12.31, Version 1.0.30(Stable)

### enhancement

- parse ast出错的时候增加文件路径

## 2013.12.23, Version 1.0.29(Stable)

### bugfix

- 修复build&combo模式下依赖错误的问题

## 2013.12.1, Version 1.0.28(Stable)

### bugfix

- 修复ignorePackageNameInUri为true时获取模块路径错误的bug

## 2013.12.1, Version 1.0.23(Stable)

### enhancement

- issue #35: 支持KISSY 1.4+的get requires from function

## 2013.11.28, Version 1.0.22(Stable)

### bugfix

- 修复对象传参数中的一个bug
- fix issue #36

## 2013.11.24, Version 1.0.21(Stable)

### enhancement

- combo增加输出文件的选项，方便fix module name
- combo、build方法支持使用对象形式传参
- fixModuleName默认为true

## 2013.11.22, Version 1.0.20(Stable)

### enhancement

- 添加对KISSY 1.3+ 新包配置（对象配置）格式的支持
- issue #34: 增加alias配置支持

## 2013.11.09, Version 1.0.19(Stable)

### enhancement

- issue #32: 解析css文件的具体信息

## 2013.11.05, Version 1.0.18(Stable)

### enhancement

- issue #24 #25: js文件解析出错时抛出异常

## 2013.10.29, Version 1.0.16(Stable)

### bugfix

- combo增加参数fixModuleName，默认为false，@issue 31

## 2013.09.23, Version 1.0.15(Stable)

### enhancement

- build增加参数traverse，可以对build的第一个参数指定的目录，递归打包，@issue 29

## 2013.08.16, Version 1.0.14(Stable)

### enhancement

- 增加了对KISSY 1.3+ ignorePkgNameInUri的支持，感谢翰文(@shepherdwind)的pull request

## 2013.07.11, Version 1.0.13(Stable)

### bugfix

- ✔ issue 23:解决部分情况下去除BOM头导致编码乱掉的问题

## 2013.07.11, Version 1.0.12(Stable)

### bugfix

- ✔ issue 23:打包的时候去除文件BOM头

## 2013.06.08, Version 1.0.11(Stable)

### bugfix

- ✔ 修正部分代码写法，combo的时候无法加模块名的bug

## 2013.06.08, Version 1.0.10(Stable)

### bugfix

- ✔ 修正require gallery模块无法被combo的问题

## 2013.05.30, Version 1.0.9(Stable)

### bugfix

- ✔ 修正require `mod/mod1`这种相对路径无法打包的问题

## 2013.05.10, Version 1.0.8(Stable)

### bugfix

- ✔ 修正combo打包时未指定charset的bug
- ✔ 修正combo打包时目录不存在的bug

## 2013.04.23, Version 1.0.7(Stable)

### bugfix

- ✔ 修正输出路径为目录导致无法正确输出的bug

## 2013.04.23, Version 1.0.6(Stable)

### bugfix

- ✔ 修正require路径为目录导致解析出错的bug
- ✔ 默认关闭任何console的log信息，silent默认为true

## 2013.04.12, Version 1.0.5(Stable)

### bugfix

- ✔ 修正部分相对路径解析出错的bug

## 2013.04.12, Version 1.0.4(Stable)

### enhancement

- ✔ 增加exclude、ignoreFiles功能支持，添加相应测试用例

## 2013.04.12, Version 1.0.3(Stable)

### enhancement

- ✔ 增加map功能支持，添加相应测试用例

## 2013.04.10, Version 1.0.2(Stable)

### enhancement

- ✔ 完全重构，增加对KISSY 1.3的自动Combo支持

## 2012.11.28, Version 0.0.6(Stable)

### enhancement

- ✔ build的时候可以直接使用入口模块名进行打包

## 2012.11.28, Version 0.0.5(Stable)

### enhancement

- ✔ 增加ignoreFiles配置，方便打包目录的时候排除一些文件

## 2012.11.28, Version 0.0.4(Stable)

### bug fix

- ✔ 修正一处self未定义bug
- ✔ 修正require正则分析少了一处引号的bug

## 2012.09.27, Version 0.0.3(Stable)

### bug fix

- ✔ #9 0.0.2中如果build没有设置编码，会输出utf-8的问题
- ✔ 判断文件是否循环过导致一些模块无法打包进来的问题

## 2012.09.19, Version 0.0.2(Stable)

### bug fix

- ✔ 修复了模块互相依赖导致死循环的问题

### enhancement
-  增加analyze接口，只分析依赖
-  build接口可以返回打包的具体信息了。
-  build接口可以单独设置输出编码了。
-  增加了silent的config项目，可以完全关闭Module Compiler的控制台输出，方便集成到自己的工具当中
-  ModuleCompiler.config可以返回当前的所有配置信息，包括包信息，方便用户查看。

## 2012.08.28, Version 0.0.1(Stable)

### bug fix

- ✔ 修复了输出编码设置无效的问题

### test case
-  增加编码测试脚本


lib/compiler.js
/**
 *
 * @author: daxingplay<daxingplay@gmail.com>
 * @time: 13-1-23 16:22
 * @description:
 */

var UglifyJS = require('uglify-js'),
    fs = require('fs'),
    path = require('path'),
    os = require('os'),
    _ = require('lodash'),
    iconv = require('iconv-lite'),
    getAst = require('./get-ast'),
    utils = require('./utils'),
    dependencies = require('./dependencies');

// modified by bachi@taobao.com 2013-12-12
// op:解析后文件（要输出的）路径
function Compiler(config, op){
    this.outputFilePath = op;
    this.modules = {};
    this.fileList = [];
    this.analyzedModules = [];
    this.combinedModules = [];
    this.buildComboModules = [];
    this.buildAnalyzedModules = [];
    this.config  = config;
    this.packages = config.packages;
}

Compiler.prototype._addModuleName = function(moduleContent, mod){
    var modName = mod.name;
    //add module path
    var start = moduleContent.indexOf('KISSY.add(');
    if(start == -1) {
        start = moduleContent.indexOf('.add(');
        if(start != -1) {
            start = start + 5;
        } else {
            start = moduleContent.indexOf('define(');
            if(start != -1){
                start += 7;
            }
        }
    } else {
        start = start + 10;
    }

    var end = moduleContent.indexOf('function', start);

    //find it
    if(start > -1 && end > start) {
        //KISSY.add(/*xxx*/function(xxx)) or KISSY.add('xxx', function(xxx))
        //remove user comments but preserve user defined module name.
        var oldModuleName = moduleContent.substring(start, end),
            moduleNameRegResult = oldModuleName.match(/^(\s*)\/\*(.*)\*\/(\s*)$/);
        if(moduleNameRegResult !== null || oldModuleName.replace(/[\r\n\s]/g, '') === ''){
            // replace user comments or white spaces with my module name.
            moduleContent = moduleContent.replace(oldModuleName, '\'' + modName + '\', ');
        }
    } else if(start > -1 && end == start) {
        //KISSY.add(function(xxx))
        var addedContentArr = [moduleContent.slice(0, start), '\'' + modName + '\','];
        if(mod.version === '1.4+' && mod.requires){
            var requires = [];
            mod.requires.forEach(function(requirePath){
                requires.push("'" + requirePath + "'");
            });
            addedContentArr.push("[" + requires.join(", ") + "], ");
        }
        addedContentArr.push(moduleContent.slice(end));
        moduleContent = addedContentArr.join('');
    }

    return moduleContent;
};
Compiler.prototype._isFileIgnored = function(filePath){
    return utils.isExcluded(filePath, this.config.ignoreFiles);
};
Compiler.prototype._isModuleExcluded = function(modName){
    return utils.isExcluded(modName, this.config.exclude);
};
Compiler.prototype._getModuleRealPathByParent = function(requiredModule, parentModule){
    var self = this;
    if(requiredModule.match(/^\./) && _.isObject(parentModule)){
        return path.resolve(path.dirname(parentModule.path), utils.addPathSuffix(requiredModule));
    }else if(!requiredModule.match(/\//) || requiredModule.match(/^gallery\//)){
        // may be kissy module.
        return requiredModule;
    }else{
        for(var i = 0; i < self.packages.length; i++){
            var pkg = self.packages[i],
                mod = requiredModule.match(new RegExp(pkg.name + '\/(.*)'));
            if(mod && mod[1]){
                var modulePath = path.resolve(pkg.path, pkg.ignorePackageNameInUri ? './' : pkg.name, utils.addPathSuffix(mod[1]));
//                console.log('%s ==> %s, package path is %s, mod1 is %s ', requiredModule, modulePath, pkg.path, utils.addPathSuffix(mod[1]));
                if(fs.existsSync(modulePath)){
                    return modulePath;
                }
            }
        }
    }
    var lastPath = path.resolve(path.dirname(parentModule.path), utils.addPathSuffix(requiredModule));
    return fs.existsSync(lastPath) ? lastPath : requiredModule;
};
Compiler.prototype._getModuleRealPathByPkg = function(filePath, pkg){
    return path.resolve(pkg.path, pkg.name, utils.addPathSuffix(filePath));
};
/**
 * get module's package by its full path.
 * @param filePath this is file full path.
 * @return {*}
 * @private
 */
Compiler.prototype._getPackage = function(filePath){
    var self = this,
        pkgPath = '',
        modPkg = null;
    if(filePath.match(/^[a-zA-Z]+$/)){
        return 'kissy';
    }
    if(filePath.match(/^gallery\//)){
        return 'gallery';
    }
    if(!fs.existsSync(filePath)){
        return null;
    }
    // TODO optimize.
    for(var i = 0; i < self.packages.length; i++){
        var pkg = self.packages[i];

        var dirname = path.dirname(filePath);
        //by hanwen 2013-08-16
        //是否属于当前包
        var isBelongPkg = false;

        if(filePath.match(new RegExp(pkg.name))){
            isBelongPkg = true;
        } else if (pkg.ignorePackageNameInUri && (dirname.indexOf(pkg.path) > -1 || path.resolve(filePath) === path.resolve(pkg.path + '.js'))) {
            //增加ignorePackageNameInUri支持，使用路径判断，如果pkg.path和
            //filePath的路径进行对比
            isBelongPkg = true;
        }

        if(isBelongPkg){

            //如果有配置ignorePackageNameInUri, 把pkg.name忽略
            var name = pkg.ignorePackageNameInUri ? '' : pkg.name;
            var curRelativePath = path.relative(path.resolve(pkg.path, name), filePath);

            if(pkgPath == '' || pkgPath.length > curRelativePath.length){
                pkgPath = curRelativePath;
                modPkg = pkg;
            }

        }
    }
    return modPkg;
};

Compiler.prototype._getModuleNameFromRequire = function(requirePath){
    var matches = /^([^\/]+?)\/\S+/.exec(requirePath);
    return matches && matches[1] ? matches[1] : '';
};
Compiler.prototype._moduleName = function(filePath, pkg){
    var relativePkgPath = path.relative(pkg.path, filePath).replace(/\\/g, '/'),
        moduleName = relativePkgPath.replace(/^\.\//, '');
    if(!utils.startWith(relativePkgPath, pkg.name + '/') && pkg.name !== 'kissy'){
        moduleName = pkg.name + '/' + moduleName;
    }
    return path.normalize(moduleName.replace(/\.js$/i, '')).replace( /\\/g, '/' );
};
Compiler.prototype._buildCombo = function(mod){
    var self = this,
        result = {};
    _.forEach(mod.dependencies, function(subMod){
        if(_.indexOf(self.buildAnalyzedModules, subMod.name) === -1){
            self.buildAnalyzedModules.push(subMod.name);
            if(subMod.pkg == 'kissy' || subMod.status != 'ok' || subMod.type === 'css'){
                self.buildComboModules.push(subMod.name);
            }
            !_.isEmpty(subMod.dependencies) && self._buildCombo(subMod);
        }
    });
    result[mod.name] = self.buildComboModules;
    return result;
};

Compiler.prototype._mapModuleName = function(modName){
    var self = this,
        rules = self.config.map;
    if(modName && !_.isEmpty(rules)){
        _.forEach(rules, function(rule){
            modName = modName.replace(new RegExp(rule[0]), rule[1]);
        });
    }
    return modName;
};
Compiler.prototype._mapContent = function(content, mod){
    var self = this,
        rules = self.config.map;
    if(content && !_.isEmpty(content)){
        _.forEach(mod.requires, function(requirePath){
            if(!/^\.{1,2}/.test(requirePath)){
                var mappedRequirePath = requirePath;
                _.forEach(rules, function(rule){
                    mappedRequirePath = requirePath.replace(new RegExp(rule[0]), rule[1]);
                });
                if(mappedRequirePath != requirePath){
                    // TODO require regex is toooooooooooo complex. So I chose the simple one.
                    content = content.replace(new RegExp("(['\"])" + requirePath + "(['\"])"), '$1' + mappedRequirePath + '$2');
                }
            }
        });
    }
    return content;
};
Compiler.prototype._generateRealModuleContent = function(mod){
    var self = this;
    if(!mod || !mod.path) return '';
    var fileContent = fs.readFileSync(mod.path);
    var modContent = iconv.decode(fileContent, mod.charset || 'utf-8');
    // fix issue 23: remove file BOM before combo.
    if(/^\uFEFF/.test(modContent)){
        modContent = modContent.toString().replace(/^\uFEFF/, '');
    }

    modContent = self._addModuleName(modContent, mod);
    return self._mapContent(modContent, mod);
};
Compiler.prototype._alias = function(modName){
    var self = this;
    if(self.config.modules && self.config.modules[modName]){
        var modConfig = self.config.modules[modName];
        if(modConfig.alias){
            var modNew = {
                name: modConfig.alias,
                path: '',
                pkg: self._getModuleNameFromRequire(modConfig.alias)
            };
            if(modNew.pkg && self.config.pkgs[modNew.pkg]){
                var modPkg = self.config.pkgs[modNew.pkg];
                var dir = modPkg.ignorePackageNameInUri ? path.resolve(modPkg.path, '../') : modPkg.path;
                modNew.path = path.resolve(dir, utils.addPathSuffix(modNew.name));
                if(fs.existsSync(modNew.path)){
                    modNew.pkg = modPkg;
                    return modNew;
                }
            }
        }
    }
    return null;
};

Compiler.prototype.getAst = getAst;

/**
 * analyze a file's dependencies.
 * @param inputFilePath please use absolute file path.
 * @return {*}
 */
Compiler.prototype.analyze = function(inputFilePath){
    var self = this,
        mod = null,
        modRequires,
        modName,
        modPkg,
        modType,
        modVersion,
        modRealPath;
    if(!inputFilePath) return null;
    // analyze module package
    modPkg = self._getPackage(inputFilePath);
    switch (modPkg){
        case null:
            !self.config.silent && console.log('cannot get package for %s, guess it is kissy module.', inputFilePath);
            mod = {
                name: inputFilePath,
                pkg: 'unknown' || self._getModuleNameFromRequire(inputFilePath),
                status: 'missing'
            };
            self.modules[mod.name] = mod;
            break;
        case 'kissy':
            mod = {
                name: inputFilePath,
                pkg: 'kissy',
                status: 'ok'
            };
            self.modules[mod.name] = mod;
            break;
        case 'gallery':
            mod = {
                name: inputFilePath,
                pkg: 'gallery',
                status: 'ok'
            };
            self.modules[mod.name] = mod;
            break;
        default :
            // get module real path based on package path.
            modRealPath = self._getModuleRealPathByPkg(inputFilePath, modPkg);
            if(!self._isFileIgnored(modRealPath)){
                modName = self._moduleName(modRealPath, modPkg);
                // map module names if user configured map rules.
                modName = self._mapModuleName(modName);
                var modAlias = self._alias(modName);
                if(modAlias){
                    modName = modAlias.name;
                    modPkg = modAlias.pkg;
                    modRealPath = modAlias.path;
                }
                if(_.indexOf(self.fileList, modRealPath) === -1){
                    // to avoid recursive analyze.
                    self.fileList.push(modRealPath);
                    if(!modRealPath.match(/\.css$/)){
                        // get this file's dependencies.
                        modRequires = dependencies.getRequiresFromFn(modRealPath, !self.config.silent);
                        if(modRequires.length){
                            modVersion = '1.4+';
                        }else{
                            modRequires = dependencies.requires(modRealPath);
                        }
                        // if user named module himself, use his name. map rules will not work then.
                        if(_.isPlainObject(modRequires) && modRequires.name){
                            modName = modRequires.name;
                            modRequires = modRequires.deps;
                        }
                        modType = 'js';
                    }else{
                        modRequires = [];
                        modType = 'css';
                    }
                    var isModExcluded = self._isModuleExcluded(modName);
                    // add this module to list.
                    mod = {
                        name: modName,
                        pkg: modPkg,
                        status: isModExcluded ? 'excluded' : 'ok',
                        path: modRealPath,
                        requires: modRequires,
                        dependencies: [],
                        dependencyTree: {},
                        charset: modPkg.charset || 'utf8',
                        type: modType,
                        version: modVersion
                    };
                    self.modules[modName] = mod;

                    if(modType === 'js'){
                        // analyze sub modules' dependencies recursively.
                        _.forEach(modRequires, function(subModPath){
                            var subMod = self.analyze(self._getModuleRealPathByParent(subModPath, mod));
                            if(subMod){
                                mod.dependencies.push(subMod);
                                mod.dependencyTree[subMod.name] = subMod.dependencyTree;
                            }
                        });

                        !isModExcluded && self.analyzedModules.push(modName);
                    }
                }else{
                    mod = self.modules[modName];
                }
            }
            break;
    }
    return mod;
};

Compiler.prototype.build = function(inputFilePath, outputFilePath, outputCharset){
    var self = this;
    var result = self.analyze(inputFilePath);

    // deal with output charset. if not specified, use charset in config.
    outputCharset = ((typeof outputCharset === 'undefined' || outputCharset === null) ? self.config.charset : outputCharset) || 'utf8';

    var combineFile = self.config.suffix ? outputFilePath.replace(/\.js$/i, self.config.suffix + '.js') : outputFilePath;
    if(path.extname(combineFile) != '.js'){
        combineFile = path.resolve(combineFile, path.basename(inputFilePath, '.js') + self.config.suffix + '.js');
    }
    result.outputFile = combineFile;

    //prepare output dir
    utils.mkdirsSync(path.dirname(combineFile));

    // if exists, unlink first, otherwise, there may be some problems with the file encoding.
    if(fs.existsSync(combineFile)){
        fs.unlinkSync(combineFile);
    }

    if(self.analyzedModules.length){
        var combinedComment = [
                '/*',
                'combined files : ',
                '',
                self.analyzedModules.join(os.EOL),
                '',
                '*/',
                ''
            ].join(os.EOL);

        var fd = fs.openSync(combineFile, 'w');
        var combinedCommentBuffer = iconv.encode(combinedComment, outputCharset || 'utf-8');
        fs.writeSync(fd, combinedCommentBuffer, 0, combinedCommentBuffer.length);
        fs.closeSync(fd);

        fd = fs.openSync(combineFile, 'a');

        _.forEach(self.analyzedModules, function(modName){
            var mod = self.modules[modName];
            // add EOL to end of file.
            var modContent = self._generateRealModuleContent(mod) + os.EOL;
            var buffer = iconv.encode(modContent, outputCharset || 'utf-8');
            fs.writeSync(fd, buffer, 0, buffer.length);
            self.combinedModules.push(modName);
        });

        fs.closeSync(fd);

    }else{
        fs.writeFileSync(combineFile, fs.readFileSync(inputFilePath));
    }

    !self.config.silent && console.info('[ok]'.bold.green+' %s ===> %s', inputFilePath, combineFile);

    result.autoCombo = self._buildCombo(result);

    result.combined = self.combinedModules;

    return result;
};
Compiler.prototype.info = function(){
    var self = this;
    var dependencies = {};
    var modules = {};
    if(_.isObject(self.modules)){
        _.forEach(self.modules, function(mod, modName){
            if(_.indexOf(self.combinedModules, modName) === -1 && !_.isEmpty(mod.dependencies)){
                var requires = [];
                _.forEach(mod.dependencies, function(subMod){
                    requires.push(subMod.name);
                });
                dependencies[modName] = {
                    requires: requires
                };
            }
            if(mod && mod.path){
                var modContent = self._generateRealModuleContent(mod);
                mod.contents = iconv.encode(modContent, mod.charset);
                modules[mod.name] = mod;
            }
        });
    }

    return {
        modules: modules,
        dependencies: dependencies
    };
}

Compiler.prototype.combo = function(fixModuleName, outputDir ,comboOnly, isJson){
    var self = this;
    var content = [];
    if(outputDir){
        outputDir = path.resolve(outputDir);
        if(!fs.existsSync(path.resolve(outputDir))){
            utils.mkdirsSync(outputDir);
        }
    }
    if(_.isObject(self.modules)){
        _.forEach(self.modules, function(mod, modName){
            if(_.indexOf(self.combinedModules, modName) === -1 && !_.isEmpty(mod.dependencies)){
                var requires = [];
                _.forEach(mod.dependencies, function(subMod){
                    requires.push(subMod.name);
                });

                if(isJson){
                    content.push("\"" + modName + "\": [\"" + requires.join("\", \"") + "\"]");
                }else{
                    content.push("'" + modName + "': { requires: ['" + requires.join("', '") + "']}");
                }
            }
            if(fixModuleName === true && mod && mod.path){
                var modContent = self._generateRealModuleContent(mod);
                var buffer = iconv.encode(modContent, mod.charset || 'utf-8');
                var outputPath = mod.path;
                if(outputDir){
                    var relativePath = mod.name;
                    if(mod.pkg && mod.pkg.ignorePackageNameInUri){
                        relativePath = mod.name.replace(mod.pkg.name + '/', '');
                    }
                    if(mod.type === 'js'){
                        relativePath = relativePath + '.js';
                    }
                    outputPath = path.resolve(outputDir, relativePath);
                    utils.mkdirsSync(path.dirname(outputPath));
                }
                // info by 拔赤
                // 建议增加开关，是否写到目标路径，这样会便于用户自定义copy规则
                // 2013-12-12
                // bugfix:grunt-kmc配置comboOnly时，文件写的位置出错的bugfix
                if(comboOnly && comboOnly === true){
                    var tModName = mod.pkg ? modName.replace(mod.pkg.name, '') : modName;
                    if(outputPath.indexOf(path.normalize(tModName)) >= 0){
                        fs.writeFileSync(outputPath, buffer);
                    }
                } else {
                    fs.writeFileSync(outputPath, buffer);
                }
            }
        });
    }
    var json="{" + os.EOL + " " + content.join("," + os.EOL + " ") + " " + os.EOL + "}";
    return content.length ?
        (isJson?json:"KISSY.config('modules', "+json+");") :
        "";
};


module.exports = Compiler;
